### ISSUE TITLE
New feature : switch config profile

### ISSUE THREAD

**Issue by azhuang** [just-containers/s6-overlay#75]
Hi, I'm considering adding a new feature to current s6-overlay init process.

The idea is to pre-deploy multiple config sets (cont-init.d, cont-finish.d and services.d) into docker image, and to switch into one profile (dev, staging etc) when launching containers with specific environment variables.

Unlike popular config management software Puppet or Chef for example, which has its own init logic thus cannot fit into s6-overlay, this feature is focused on profile switching only. You still have to somehow generate config files, while it helps when you need to change container configs to fit variety environments.

I have updated my fork at https://github.com/just-containers/s6-overlay/compare/master...azhuang:profile-switch .

I'm using S6_CONTAINER_PROFILE to pass profile name into container. To be compatible with previous version, it's default to null so that profile switching is disabled.

Any profile has its own directory at /etc/cont-profile.d/<profile_name>/ . Each of the profile currently has 3 sub directories as cont-init.d, cont-finish.d and services.d . The active profile will be automatically (symbolic) linked to /etc/{cont-init.d, cont-finish.d, services.d}.

The switching process is executed after fix-attr.d so that you can safely update container profile by mounting a volume into /etc/cont-profile.d/ . Then the init process will continue to load init scripts and services.

My work has not been tested thoroughly. I'd like to know what you think about this feature at first.

**Comment by jprjr** [comment]
@azhuang I think this is a pretty neat idea. I could see this being useful to maybe turn off debugging-type features.

**Comment by glerchundi** [comment]
@azhuang first of all, thanks for contributing :-)

I think this is a very good idea but I'm not pretty sure if this is necessary to include in our overlay. There are several ways to achieve the same behaviour:

1. Create one container per profile: myimage:v1-dev, myimage:v1-staging, myimage:v1-production
2. Use one image but make different folders available in each execution. For example, `docker run -v ``pwd``/dev/cont-init.d:/etc/cont-init.d myimage:v1`, `docker run -v ``pwd``/staging/cont-init.d:/etc/cont-init.d myimage:v1` or , `docker run -v ``pwd``/prod/cont-init.d:/etc/cont-init.d myimage:v1`
3. Your option, check it in runtime and use them accordingly.

I would bet for the second choice because:
* It doesn't overcomplicate the init process
* Better granularity. It allows you to override just one of them.
* It keeps everything _simple_ for the end-user.

But this would require removing `00-runscripts` from `/etc/fix-attrs.d` and embed its content inside the init process, as it is part of the system-wide init process.

WDYT?

**Comment by azhuang** [comment]
@jprjr @glerchundi 

Well your inputs reminds me something a little off-topic. We really should put all user configurables into one directory so that we can bind mount all scripts at once. Currently we can't mount `/etc` because `mount` behaves like it will replace all contents from its mounting point. Besides user scripts in `cont-init.d` `cont-finish.d` `services.d`, `00-runscripts` should be included as well.

Yes I think `00-runscripts` should be part of init process. It is vital to overlay and end-user may override this file by mistake, especially when it is mounted into container along with other user scripts.

My implementation is actually not a must but an option. If `S6-CONTAINER-PROFILE` is not set, the switching process will be skipped (when original `/etc/cont-init.d` will be used). It is backward compatible so that end-user can update to new version without modifying configs. BTW, one profile doesn't have to include `cont-init.d` `cont-finish.d`. The `services.d` is a bit tricky, s6-svscan will report error if `services.d` is symlinked to a directory that doesn't exist. (I have made a simple fix, but I'd like to rewrite it according to how we organize user scripts.)

**Comment by glerchundi** [comment]
> Well your inputs reminds me something a little off-topic. We really should put all user configurables into one directory so that we can bind mount all scripts at once. Currently we can't mount /etc because mount behaves like it will replace all contents from its mounting point.

Not necessarily:

This would share services and will mount just the required ones for each environment:

**dev** (hook init and finish):
```
docker run \
  -v `pwd`/dev/cont-init.d:/etc/cont-init.d \
  -v `pwd`/dev/cont-finish.d:/etc/cont-finish.d \
  myimage:v1
```

**prod** (do nothing probably it already contains everything production-ready):
```
docker run myimage:v1
```

> Besides user scripts in cont-init.d cont-finish.d services.d, 00-runscripts should be included as well.
> 
> Yes I think 00-runscripts should be part of init process. It is vital to overlay and end-user may override this file by mistake, especially when it is mounted into container along with other user scripts.

Yep. Done in ... .

> My implementation is actually not a must but an option. If S6-CONTAINER-PROFILE is not set, the switching process will be skipped (when original /etc/cont-init.d will be used).

Right, but it allows you to do the same in so different ways that it can be confusing.

> It is backward compatible so that end-user can update to new version without modifying configs.

Of course and it's the right way to do the things :-)

> BTW, one profile doesn't have to include cont-init.d cont-finish.d. 

But now sharing between profiles is a little bit more complicated:
```
/etc/cont-profile.d/base
  - /cont-init.d
  - /cont-finish.d
  - /services.d
/etc/cont-profile.d/dev
 - /cont-init.d
 - /cont-finish.d
 - /services.d -> /etc/cont-profile.d/base/services.d
/etc/cont-profile.d/prod
 - /cont-init.d -> /etc/cont-profile.d/base/cont-init.d
 - /cont-finish.d -> /etc/cont-profile.d/base/cont-finish.d
 - /services.d -> /etc/cont-profile.d/base/services.d
[...]
```

Or something like this.

> The services.d is a bit tricky, s6-svscan will report error if services.d is symlinked to a directory that doesn't exist. (I have made a simple fix, but I'd like to rewrite it according to how we organize user scripts.)

It's hiercopied to `/var/run` first and then `s6-svscan` is used there so it shouldn't be a problem if it's its probably because it's a bug :-)

That being said, we can include this behaviour inside the init process without promoting it anywhere because I'm not quite sure what is my opinion on this. I feel like I'm violating some rules on lean containers, like deviating from what people is doing right now.

Create a rebased PR and we'll discuss there.

**Comment by glerchundi** [comment]
Although I mistakenly closed the #81 PR, it's already pushed into the master. And a new version is on the way: v1.14.0.3.
